File: QE-Java Selenium/Activity21.java
package demos;

import java.time.Duration;
import java.util.Set;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.firefox.FirefoxDriver;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

public class Activity21 {

    public static void main(String[] args) {

        WebDriver driver = new FirefoxDriver();
        driver.manage().window().maximize();

        WebDriverWait wait = new WebDriverWait(driver, Duration.ofSeconds(10));

        driver.get("https://training-support.net/webelements/tabs");

        System.out.println("Main Page Title: " + driver.getTitle());

        String parentHandle = driver.getWindowHandle();
        System.out.println("Parent Handle: " + parentHandle);

        String openTabXpath = "//button[text()='Open A New Tab']";

        driver.findElement(By.xpath(openTabXpath)).click();

        wait.until(ExpectedConditions.numberOfWindowsToBe(2));

        Set<String> handles = driver.getWindowHandles();
        for (String handle : handles) {
            if (!handle.equals(parentHandle)) {
                driver.switchTo().window(handle);
                break;
            }
        }

        System.out.println("First New Tab Title: " + driver.getTitle());
        String message = wait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath("/html/body/div/main/div/div/div/div/h2"))).getText();

        System.out.println("First New Tab Message: " + message);

        driver.quit();
    }
}

File: QE-Python Selenium/Activity21.py
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
import time

driver = webdriver.Firefox()
driver.maximize_window()

wait = WebDriverWait(driver, 10)

driver.get("https://training-support.net/webelements/tabs")

print("Main Page Title:", driver.title)

parent_handle = driver.current_window_handle
print("Parent Handle:", parent_handle)

open_tab_xpath = "//button[text()='Open A New Tab']"

driver.find_element(By.XPATH, open_tab_xpath).click()

wait.until(EC.number_of_windows_to_be(2))

for handle in driver.window_handles:
    if handle != parent_handle:
        driver.switch_to.window(handle)
        break

print("First New Tab Title:", driver.title)

message = wait.until(EC.visibility_of_element_located((By.XPATH, "/html/body/div/main/div/div/div/div/h2"))).text

print("First New Tab Message:", message)

time.sleep(2)
driver.quit()
